### PR TITLE
Custom war test deploy

### PR DESCRIPTION
**Pull request by kuisathaverat**
- Detect PR on the repo
- build custom war
- build docker image
- deploy to k8s
- run sample jobs and check results

follow up of https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/20

### CODE REVIEW

**Comment by kuisathaverat** [comment]
waiting for jenkinsci/custom-war-packager#33

**Comment by kuisathaverat** [comment]
Finally done and working on jenkins.arc

File: src/test/it/cdkata.yml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    war:
+      groupId: "org.jenkins-ci.main"
+      artifactId: "jenkins-war"
+      source:
+        version: 2.121.1
+    plugins:     
+      - groupId: "org.jenkins-ci.plugins.workflow"
+        artifactId: "workflow-api"
+        source:
+          version:  2.28
```
Comment (jglick) [inline src/test/it/cdkata.yml]: This is already wrong. See #73.

IMO we should not merge anything that duplicates configuration specified in `pom.xml`.

File: src/test/it/cdkata.yml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        source:
+          version: 2.3
+      - groupId: "io.jenkins.plugins"
+        artifactId: "aws-global-configuration"
+        source:
+          version: 1.0
+      - groupId: "io.jenkins.plugins"
+        artifactId: "artifact-manager-s3"
+        source:
+          version: 1.1
```
Comment (jglick) [inline src/test/it/cdkata.yml]: Why is this here?

File: src/test/it/cdkata.yml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          dir: ../../..
+    systemProperties: {
```
Comment (jglick) [inline src/test/it/cdkata.yml]: Why the unusual YAML syntax?

File: src/test/it/cdkata.yml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  tests:
+    - "core.ArtifactsTest"
+#    - "plugins.CopyArtifactPluginTest"
+#    - "plugins.CompressArtifactsPluginTest"
+#    - "plugins.WorkflowPluginTest"
+#    - "plugins.WorkflowMultibranchTest"
+  categories:
+    - "org.jenkinsci.test.acceptance.junit.SmokeTest"
+pct:
+  disabled: true
```
Comment (jglick) [inline src/test/it/cdkata.yml]: So…what is the point of running this at all then? My only real interest in this kind of tool is to look for PCT failures proactively.

File: src/test/it/cdkata.yml [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  useLocalSnapshots: false
+  jth:
+    version: 2.38
```
Comment (jglick) [inline src/test/it/cdkata.yml]: Should be omitted. Every plugin repo defines a compatible version of JTH and this should not be overridden.

File: src/test/it/initScripts/01-enableArtifactManager.groovy [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import io.jenkins.plugins.artifact_manager_jclouds.s3.S3BlobStoreConfig
+import jenkins.model.ArtifactManagerConfiguration
+import io.jenkins.plugins.artifact_manager_jclouds.JCloudsArtifactManagerFactory
+import io.jenkins.plugins.artifact_manager_jclouds.s3.S3BlobStore
+import io.jenkins.plugins.aws.global_configuration.CredentialsAwsGlobalConfiguration
+
+def factory = new JCloudsArtifactManagerFactory(new S3BlobStore())
+println("--- Enabling default artifact storage: ${factory.descriptor.displayName}")
+
+ArtifactManagerConfiguration.get().artifactManagerFactories.add(factory)
```
Comment (jglick) [inline src/test/it/initScripts/01-enableArtifactManager.groovy]: We have JEP-201 compatibility now, so why would we not use it?

File: src/test/it/initScripts/99-saveToDisk.groovy [hunk 7]
```diff
@@ -0,0 +1,3 @@
+import jenkins.model.Jenkins
+
+Jenkins.instance.save()
```
Comment (jglick) [inline src/test/it/initScripts/99-saveToDisk.groovy]: Why is this necessary? Every setter to a `Saveable` is expected to save the object when modified.

File: src/test/it/jenkins.yml [hunk 8]
```diff
@@ -0,0 +1,36 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    iam.amazonaws.com/role: <IAM_ROLE>
+  generateName: <NAME>-
+  labels:
+    name: jnlp
```
Comment (jglick) [inline src/test/it/jenkins.yml]: The filename `jenkins.yml` implies that this is a pod for the master, yet you seem to be defining an agent here?

File: src/test/it/jenkins.yml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  - name: <NAME>
+    image: <DOCKER_IMAGE>
+    tty: true
+    env:
+      - name: JAVA_OPTS
+        value: -Djava.awt.headless=true
+      - name: S3_REGION
+        value: <S3_REGION>
+      - name: S3_CONTAINER
+        value: <S3_CONTAINER>
```
Comment (jglick) [inline src/test/it/jenkins.yml]: Why are these things being defined here? They are covered by regular settings.

File: src/test/it/runJobs.sh [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  echo $(curl -s -o /dev/null -w '%{http_code}' http://127.0.0.1:8080/job/${job}/build?delay=0sec)
+}
+
+waitFinishJob(){
+  local job=${1:-?}
+  RUNNING="<building>true</building>"
+  while [ "${RUNNING}" = "<building>true</building>" ] 
+  do
+    sleep 10
+    RUNNING="$(curl -sS http://127.0.0.1:8080/job/${job}/1/api/xml?xpath=/workflowRun/building)"
```
Comment (jglick) [inline src/test/it/runJobs.sh]: Use `tree` in addition to `xpath`.

File: src/test/it/runJobs.sh [hunk 11]
```diff
@@ -0,0 +1,85 @@
+#!/bin/sh
+buildJob(){
+  local job=${1:-?}
+  echo $(curl -s -o /dev/null -w '%{http_code}' http://127.0.0.1:8080/job/${job}/build?delay=0sec)
+}
+
+waitFinishJob(){
+  local job=${1:-?}
+  RUNNING="<building>true</building>"
+  while [ "${RUNNING}" = "<building>true</building>" ]
```
Comment (jglick) [inline src/test/it/runJobs.sh]: A better `xpath` (`…/text()` IIRC) would strip the tags.

File: src/test/it/runJobs.sh [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+getDurationJob(){
+  local job=${1:-?}
+  echo "$(curl -sS http://127.0.0.1:8080/job/${job}/1/api/xml?xpath=/workflowRun/duration)"
+}
+
+deleteJob(){
+  local job=${1:-?}
+  echo "$(curl -s -o /dev/null -w '%{http_code}' http://127.0.0.1:8080/job/${job}/doDelete)"
```
Comment (jglick) [inline src/test/it/runJobs.sh]: You must POST to this. In fact you _are_ getting a 405 response—but the test passes anyway. So this `curl` usage is not actually verifying anything AFAICT.

File: src/test/it/jenkins.yml (hunk 8 above)
Comment (kuisathaverat) [reply src/test/it/jenkins.yml]: this file is used in the pipeline library to provision a pod

File: src/test/it/initScripts/99-saveToDisk.groovy (hunk 7 above)
Comment (kuisathaverat) [reply src/test/it/initScripts/99-saveToDisk.groovy]: C&P from CWP demo scripts

File: src/test/it/cdkata.yml (hunk 4 above)
Comment (kuisathaverat) [reply src/test/it/cdkata.yml]: at the moment of this PR was not ready, @raul-arabaolaza make some progress on that

File: src/test/it/runJobs.sh (hunk 12 above)
Comment (kuisathaverat) [reply src/test/it/runJobs.sh]: the only not tested results are for doDelete operations, all the rest check the return test or code.

File: src/test/it/runJobs.sh (hunk 10 above)
Comment (kuisathaverat) [reply src/test/it/runJobs.sh]: the job is selected on the URL
